src/librustc_typeck/check/demand.rs

```diff
@@ -239,6 +239,19 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                 }
                 None
             }
+            (expected_ty, &ty::TyRef(_, checked_ty)) => {
+                if expected_ty == &checked_ty.ty.sty {
+                    let sp = self.sess().codemap().call_span_if_macro(expr.span);
+                    if let Ok(src) = self.tcx.sess.codemap().span_to_snippet(sp) {
+                        return if src.starts_with("&") {
```

Comment:
You should probably destructure `expr` to figure out if it's an `ExprRef`. This way you can obtain the inner item directly. Your current code would suggest `mut String::new()` for `foo(&mut String::new())`.